CASSANDRA-13630 support large internode messages with netty

jasobrown opened the pull request.

jasobrown committed:
Support sanely sending large messages with internode messaging via netty

jasobrown committed:
clean up error handling logic in MIH; allow custom timeouts for blocking on the queue in RebufferingByteBufDataInputPlus; fix MIHTest

jasobrown committed:
add RebufferingByteBufDataInputPlus timeout test

jasobrown committed:
fix ChannelWriterTest to wait for asynchronous blokcing implemenations

jasobrown committed:
point to jasobrown dtest repo

jasobrown committed:
trivial lambda change

dineshjoshi commented on src/java/org/apache/cassandra/net/async/MessageOutHandler.java:
```diff
@@ -140,8 +125,17 @@ public void write(ChannelHandlerContext ctx, Object o, ChannelPromise promise)
+            // if we allocated too much buffer for this message, we'll log here.
+            // if we allocated to little buffer space, we would have hit an exception when trying to write more bytes to it
+            if (out.isWritable())
+                errorLogger.error("{} reported message size {}, actual message size {}, msg {}",
```
How likely are we to cause log spam?

jasobrown replied:
The `errorLogger` is an instance of `NoSpamLogger`, so we can adjust the log intervals. I'm not sure if making it an assert is better (as that would throw and Exception, and the connection would get killed). As it's mostly a developer-level logging thing, I could just remove it, as well.  wdyt?

dineshjoshi replied:
It's fine. We can leave it in.

dineshjoshi commented on src/java/org/apache/cassandra/net/async/RebufferingByteBufDataInputPlus.java:
```diff
@@ -36,6 +37,11 @@
+    /**
+     * Default to a very large value.
+     */
+    private static final long DEFAULT_REBUFFER_BLOCK_IN_MILLIS = TimeUnit.DAYS.toMillis(2);
```
Its risky to default to a very large value. Why do we want to default to such a large value?

jasobrown replied:
The corresponding pre-4.0 code just does blocking IO on a socket, with no timeouts. Thus it blocks forever. 2 days seems shorter than forever, but I'm game to change this to any time-bounded value.

dineshjoshi replied:
I think it doesn't make sense to wait for more than 5 minutes to rebuffer. Given buffer sizes are on the order of a few megabytes at most, 5 minutes is an eternity. It's best to keep it short so failures are exposed sooner than later.

jasobrown replied:
I've lowered it to 3 minutes.

dineshjoshi commented on src/java/org/apache/cassandra/net/async/RebufferingByteBufDataInputPlus.java:
```diff
@@ -183,6 +195,11 @@ public int available() throws EOFException
 
+    public boolean isEmpty() throws EOFException
+    {
+        return available() == 0;
```
The method `available()` has a side effect of `channelConfig.setAutoRead(true)` when the `availableBytes` falls below the `lowWatermark`. Are you sure that is ok?

jasobrown replied:
Nice find. I added that side effect to 
[StreamingInboundHandler to help deserialization](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/streaming/async/StreamingInboundHandler.java#L167). Clearly, I could move that into a separate method, and have `StreamingInboundHandler` invoke that after the `available()` call. wdyt?

jasobrown replied:
So, looking at this again, I needed to figure out why `available()` does the autoRead check. For posterity, here's the reasoning behind it:

There a case where we enqueue a buffer, and disable autoRead. Then read all of the buffer exactly (and no more bytes are available to read in the instance). In an effort to not block indefinitely (so I could check if `StreamingInboundHandler.close` is true), i called available() (and did the autoRead check) instead of attempt to read a value from the inputPlus (which would call `reBuffer()` and block either indefinitely or throw an exception if I added a timeout). I didn't want the indefinite block nor an exception to be thrown, so I bolted the autoRead check onto `available()`.

I think it still makes sense to move this into a separate method call.

dineshjoshi commented on src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java:
```diff
@@ -393,6 +393,18 @@ private Channel getOrCreateChannel()
+        {
+            try
+            {
+                session.onError(t).get(5, TimeUnit.MINUTES);
```
Why do we have a 5 minute timeout? We should pull this out as a constant.

jasobrown replied:
done

dineshjoshi commented on src/java/org/apache/cassandra/net/MessageIn.java:
```diff
@@ -231,4 +241,437 @@ public String toString()
+        }
+
+        @SuppressWarnings("resource")
+        public void process(ByteBuf in) throws IOException
```
This method is a bit complicated to understand. Could you refactor it? For example`RedisDecoder` provides a nice example of this - https://github.com/netty/netty/blob/4.1/codec-redis/src/main/java/io/netty/handler/codec/redis/RedisDecoder.java

jasobrown replied:
done. I also pulled the main loop logic into the base class.

jasobrown committed:
moved autoRead check out of available method and into it's own method; added tests

jasobrown committed:
refactored MessageInProcessor.process to move the main loop logic into the base class, and moved logic for each case statement into submethods rather than directly in-line with the loop

dineshjoshi commented on src/java/org/apache/cassandra/net/MessageIn.java:
```diff
@@ -231,4 +241,437 @@ public String toString()
+            {
+                String key = DataInputStream.readUTF(in);
+                ParameterType parameterType = ParameterType.byName.get(key);
+                int valueLength = in.readInt();
```
Unused variable.

jasobrown replied:
removed

dineshjoshi commented on src/java/org/apache/cassandra/net/MessageIn.java:
```diff
@@ -231,4 +241,437 @@ public String toString()
+            {
+                String key = DataInputStream.readUTF(inputTracker);
+                ParameterType parameterType = ParameterType.byName.get(key);
+                long valueLength = VIntCoding.readUnsignedVInt(inputTracker);
```
Unused variable.

dineshjoshi commented on src/java/org/apache/cassandra/net/async/ByteBufDataOutputStreamPlus.java:
```diff
@@ -196,10 +220,12 @@ protected void doFlush(int count) throws IOException
             currentBuf.writerIndex(byteCount);
 
-            if (!Uninterruptibles.tryAcquireUninterruptibly(channelRateLimiter, byteCount, 2, TimeUnit.MINUTES))
+            if (!Uninterruptibles.tryAcquireUninterruptibly(channelRateLimiter, byteCount, rateLimiterBlockTime, rateLimiterBlockTimeUnit))
```
This is going up from 2 minutes to potentially 5 minutes. I think this should be ok but just making sure.

jasobrown replied:
yes, i intentionally set it to 2 minutes.

jasobrown replied:
Good catch. ftr, the only place where this went from 2 up to 5 minutes was in `NettyStreamingMessageSender.FIleSendTask` where I [defaulted to 5 minutes](https://github.com/apache/cassandra/pull/253/files#diff-6d6dd6c9e52acdc8462538a0d3695c0aR327). I've switched that call site to default to 2 minutes.

dineshjoshi commented on test/unit/org/apache/cassandra/net/MessageInProcessorPre40Test.java:
```diff
@@ -0,0 +1,115 @@
+
+    private void buildParamBufPre40(int valueLength) throws IOException
+    {
+        buf = Unpooled.buffer(1024, 1024); // 1k should be enough for everybody!
```
🤣

dineshjoshi commented on src/java/org/apache/cassandra/net/async/MessageInHandler.java:
```diff
@@ -18,143 +18,296 @@
      */
-    @SuppressWarnings("resource")
-    public void handleDecode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception
+    class NonblockingBufferHandler implements BufferHandler
```
This can be made into a static inner class.

jasobrown replied:
done

jasobrown committed:
changed a few timeout values

jasobrown committed:
fix eclipse-warning

dineshjoshi commented on src/java/org/apache/cassandra/net/MessageOut.java:
```diff
@@ -180,6 +199,73 @@ public String toString()
+    {
+        for (int ii = 0; ii < parameters.size(); ii += PARAMETER_TUPLE_SIZE)
+        {
+            if (((ParameterType)parameters.get(ii + PARAMETER_TUPLE_TYPE_OFFSET)).equals(type))
```
Don't need the typecast to `ParameterType`

jasobrown replied:
done

dineshjoshi commented on src/java/org/apache/cassandra/net/MessageOut.java:
```diff
@@ -180,6 +199,73 @@ public String toString()
+    {
+        try
+        {
+            UUID sessionId =  (UUID)getParameter(ParameterType.TRACE_SESSION);
```
Nit: Spacing

jasobrown replied:
fixed

dineshjoshi commented on src/java/org/apache/cassandra/net/MessageOut.java:
```diff
@@ -180,6 +199,73 @@ public String toString()
+        try
+        {
+            UUID sessionId =  (UUID)getParameter(ParameterType.TRACE_SESSION);
+            if (sessionId != null)
```
I assume `sessionId != null` means that tracing is enabled? Otherwise we should explicitly check whether tracing is enabled.

jasobrown replied:
Yes. Further, this method is basically taken verbatim from 3.11's [OutboundTcpConnection.writeConnected()](https://github.com/apache/cassandra/blob/cassandra-3.11/src/java/org/apache/cassandra/net/OutboundTcpConnection.java#L312) method.

jasobrown committed:
address Dinesh review comments

dineshjoshi commented on test/unit/org/apache/cassandra/net/async/ChannelWriterTest.java:
```diff
@@ -70,32 +86,71 @@ public static void before()
+        if (type == ChannelWriterType.COALESCING)
+            coalescingStrategy = CoalescingStrategies.newCoalescingStrategy(CoalescingStrategies.Strategy.FIXED.name(), COALESCE_WINDOW_MS, null, "test");
+        else
+            coalescingStrategy = Optional.empty();
```
This is beyond the scope of this PR, however, we should refactor `OutboundConnectionParams` to not store an `Optional`.

jasobrown replied:
Wait for CASSANDRA-14503 :)

jasobrown commented:
Wait for CASSANDRA-14503 :)

On Sun, Sep 9, 2018 at 18:37 Dinesh <notifications@github.com> wrote:

> *@dineshjoshi* commented on this pull request.
> ------------------------------
>
> In test/unit/org/apache/cassandra/net/async/ChannelWriterTest.java
> <https://github.com/apache/cassandra/pull/253#discussion_r216180830>:
>
> >          channel = new EmbeddedChannel();
>          omc = new NonSendingOutboundMessagingConnection(id, null, Optional.empty());
> -        channelWriter = ChannelWriter.create(channel, omc::handleMessageResult, Optional.empty());
> -        channel.pipeline().addFirst(new MessageOutHandler(id, MessagingService.current_version, channelWriter, () -> null));
> -        coalescingStrategy = CoalescingStrategies.newCoalescingStrategy(CoalescingStrategies.Strategy.FIXED.name(), COALESCE_WINDOW_MS, null, "test");
> +        wrapper = new MessageOutWrapper();
> +        OutboundConnectionParams.Builder builder = OutboundConnectionParams.builder()
> +                                                                           .messageResultConsumer(this::handleMessageResult)
> +                                                                           .coalescingStrategy(Optional.empty())
> +                                                                           .protocolVersion(MessagingService.current_version)
> +                                                                           .connectionId(id);
> +
> +        if (type == ChannelWriterType.COALESCING)
> +            coalescingStrategy = CoalescingStrategies.newCoalescingStrategy(CoalescingStrategies.Strategy.FIXED.name(), COALESCE_WINDOW_MS, null, "test");
> +        else
> +            coalescingStrategy = Optional.empty();
>
> This is beyond the scope of this PR, however, we should refactor
> OutboundConnectionParams to not store an Optional.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/cassandra/pull/253#pullrequestreview-153604787>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABVYQazZAIsQIg5RPNy6jGI5KELiH5Cdks5uZcJYgaJpZM4WGkHi>
> .
>

smiklosovic commented:
@jasobrown can be this PR closed?

pmcfadin commented:
This issue was resolved and merged separately

https://issues.apache.org/jira/browse/CASSANDRA-13630
